File: app/src/main/java/org/ole/planet/myplanet/data/NetworkResult.kt
package org.ole.planet.myplanet.data

sealed class NetworkResult<out T> {
    data class Success<out T>(val data: T) : NetworkResult<T>()
    data class Error(val code: Int?, val message: String?) : NetworkResult<Nothing>()
    data class Exception(val exception: Throwable) : NetworkResult<Nothing>()
}

File: flutter/lib/core/network/network_result.dart
/// Port of `data/NetworkResult.kt`.
///
/// Kotlin's `sealed class NetworkResult<out T>` becomes a Dart sealed class, so
/// `switch` over it stays exhaustive at compile time the same way `when` did.
sealed class NetworkResult<T> {
  const NetworkResult();
}

class NetworkSuccess<T> extends NetworkResult<T> {
  const NetworkSuccess(this.data);

  final T data;
}

/// A response that arrived but was not successful (non-2xx).
class NetworkError<T> extends NetworkResult<T> {
  const NetworkError(this.code, this.message);

  final int? code;
  final String? message;
}

/// A request that never produced a response (socket error, timeout, parse failure).
class NetworkException<T> extends NetworkResult<T> {
  const NetworkException(this.error);

  final Object error;
}
